pkg/storage/internal/sync/sync_chunk_action.go

```diff
@@ -0,0 +1,130 @@
+package sync
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"iter"
+	"log/slog"
+
+	"github.com/4chain-ag/go-wallet-toolbox/pkg/internal/logging"
+	"github.com/4chain-ag/go-wallet-toolbox/pkg/wdk"
+	"github.com/go-softwarelab/common/pkg/must"
+	"github.com/go-softwarelab/common/pkg/seq"
+	"github.com/go-softwarelab/common/pkg/to"
+)
+
+const (
+	maxIterations = 1000
+	minPageSize   = 10
+)
+
+type syncChunkAction struct {
+	logger   *slog.Logger
+	repo     Repository
+	chunkers []Chunker
+}
+
+func newSyncChunkAction(logger *slog.Logger, repo Repository) *syncChunkAction {
+	return &syncChunkAction{
+		logger:   logging.Child(logger, "getSyncChunk"),
+		repo:     repo,
+		chunkers: all(repo),
+	}
+}
+
+func (s *syncChunkAction) GetSyncChunk(ctx context.Context, args *wdk.RequestSyncChunkArgs) (*wdk.SyncChunk, error) {
+	user, err := s.repo.FindUser(ctx, args.IdentityKey)
+	if err != nil {
+		return nil, fmt.Errorf("cannot find user: %w", err)
+	}
+
+	if user == nil {
+		return nil, fmt.Errorf("user with identity key %s not found", args.IdentityKey)
+	}
+
+	chunk := &wdk.SyncChunk{
+		FromStorageIdentityKey: args.FromStorageIdentityKey,
+		ToStorageIdentityKey:   args.ToStorageIdentityKey,
+		UserIdentityKey:        args.IdentityKey,
+	}
+
+	if args.Since == nil || user.UpdatedAt.After(*args.Since) {
+		chunk.User = user
+	}
+
+	if err = s.process(ctx, args, user.UserID, chunk); err != nil {
+		return nil, fmt.Errorf("failed to process sync chunk: %w", err)
+	}
+
+	return chunk, nil
+}
+
+func (s *syncChunkAction) process(ctx context.Context, args *wdk.RequestSyncChunkArgs, userID int, result *wdk.SyncChunk) error {
+	var err error
+	var itemsCounter uint64
+
+	offsetsLookup := s.makeOffsetsLookup(args)
+
+	for _, chunker := range s.chunkers {
+		if !chunker.IsApplicable(offsetsLookup) {
+			continue
+		}
+
+		var relativeOffset uint64
+
+		for range s.safeLoopIterator(maxIterations) {
+			if err = ctx.Err(); err != nil {
+				return fmt.Errorf("context canceled, aborting: %w", err)
+			}
+
+			freeSlots := args.MaxItems - itemsCounter
+			limit, err := to.UInt64(min(freeSlots, max(minPageSize, args.MaxItems/chunker.MaxDivider())))
+			if err != nil {
+				return fmt.Errorf("failed to calculate limit: %w", err)
+			}
```

Comment:
I belive that this could be done that way with clearer intention:
```suggestion
			limit := to.ValueBetween(args.MaxItems/chunker.MaxDivider(), minPageSize, freeSlots)
```

Reply:
I merged your suggestion, so you already addressed that 🙈